### PR TITLE
parser: implement Restore for TableSource and Join

### PR DESCRIPTION
**Pull request by exialin**
Implement `Restore` and unit tests for `TableSource` and `Join`.

The `Source` of `TableSource` can be a `SelectStmt` or a `UnionStmt`, which haven't been implemented yet, so they are not covered in the test.

---
Issue: https://github.com/pingcap/tidb/issues/8532

### CODE REVIEW

File: ast/dml_test.go [hunk 1]
```diff
@@ -139,3 +139,32 @@ func (tc *testDMLSuite) TestLimitRestore(c *C) {
...
+	RunNodeRestoreTest(c, testCases, "select * from %s", extractNodeFunc)
+}
+
+func (tc *testDMLSuite) TestJoinRestore(c *C) {
+	testCases := []NodeRestoreTestCase{
+		{"t1 natural join t2", "`t1` NATURAL JOIN `t2`"},
+		{"t1 natural left join t2", "`t1` NATURAL LEFT JOIN `t2`"},
+		{"t1 natural right outer join t2", "`t1` NATURAL RIGHT JOIN `t2`"},
+		{"t1 straight_join t2", "`t1` STRAIGHT_JOIN `t2`"},
+		{"t1 straight_join t2 on t1.a=t2.a", "`t1` STRAIGHT_JOIN `t2` ON `t1`.`a`=`t2`.`a`"},
```
Comment (kennytm) [inline ast/dml_test.go]: Perhaps use an operator which allows us not to rely on #106?

File: ast/dml.go [hunk 2]
```diff
@@ -85,7 +85,51 @@ type Join struct {
 
 // Restore implements Node interface.
 func (n *Join) Restore(ctx *RestoreCtx) error {
-	return errors.New("Not implemented")
+	if err := n.Left.Restore(ctx); err != nil {
+		return errors.Annotate(err, "An error occurred while restore Join.Left")
+	}
+	ctx.WritePlain(" ")
```
Comment (kennytm) [inline ast/dml.go]: You could skip this extra call if you attach the space to the `JOIN`/`STRAIGHT_JOIN` keyword.

```go
if condition {
    ctx.WriteKeyWord(" NATURAL")
}
if condition {
    ctx.WriteKeyWord(" LEFT")
}
ctx.WriteKeyWord(" JOIN ")
```

File: ast/dml_test.go [hunk 3]
```diff
@@ -139,3 +139,32 @@ func (tc *testDMLSuite) TestLimitRestore(c *C) {
...
+		{"t1 cross join t2", "`t1` JOIN `t2`"},
+		{"t1 cross join t2 on t1.a=t2.a", "`t1` JOIN `t2` ON `t1`.`a`=`t2`.`a`"},
+		{"t1 join t2 using (b)", "`t1` JOIN `t2` USING (`b`)"},
+		{"t1 join t2 using (b,c) left join t3 on t1.a=t3.a", "`t1` JOIN `t2` USING (`b`,`c`) LEFT JOIN `t3` ON `t1`.`a`=`t3`.`a`"},
+		{"t1 natural join t2 right join t3 using (b,c)", "`t1` NATURAL JOIN `t2` RIGHT JOIN `t3` USING (`b`,`c`)"},
```
Comment (kennytm) [inline ast/dml_test.go]: Please add a test case involving `INNER JOIN`.

File: opcode/opcode.go [hunk 4]
```diff
@@ -109,7 +109,7 @@ var opsLiteral = map[Op]string{
 	RightShift: ">>",
 	GE:         ">=",
 	LE:         "<=",
-	EQ:         "==",
+	EQ:         "=",
```
Comment (kennytm) [inline opcode/opcode.go]: (This should stay as part of #106, not in this PR.)

File: ast/dml_test.go [hunk 5]
```diff
@@ -139,3 +139,34 @@ func (tc *testDMLSuite) TestLimitRestore(c *C) {
...
+		{"t1 natural right outer join t2", "`t1` NATURAL RIGHT JOIN `t2`"},
+		{"t1 straight_join t2", "`t1` STRAIGHT_JOIN `t2`"},
+		{"t1 straight_join t2 on t1.a>t2.a", "`t1` STRAIGHT_JOIN `t2` ON `t1`.`a`>`t2`.`a`"},
+		{"t1 cross join t2", "`t1` JOIN `t2`"},
+		{"t1 cross join t2 on t1.a>t2.a", "`t1` JOIN `t2` ON `t1`.`a`>`t2`.`a`"},
+		{"t1 inner join t2 using (b)", "`t1` JOIN `t2` USING (`b`)"},
+		{"t1 join t2 using (b,c) left join t3 on t1.a>t3.a", "`t1` JOIN `t2` USING (`b`,`c`) LEFT JOIN `t3` ON `t1`.`a`>`t3`.`a`"},
+		{"t1 natural join t2 right join t3 using (b,c)", "`t1` NATURAL JOIN `t2` RIGHT JOIN `t3` USING (`b`,`c`)"},
+		// {"t1, t2", "`t1` JOIN `t2`"},
+		// {"t1, t2, t3", "`t1` JOIN `t2` JOIN `t3`"},
```
[the remark this answers is not in the document]
Comment (exialin) [reply ast/dml_test.go]: @kennytm  Oh I didn't consider the parentheses. The above test case is not handled correctly. I'll fix it.
 By the way, the AST of `select * from t1, t2` is the same as that of `select * from (t1) join t2`.

**Comment by exialin** [comment]
@kennytm I added a variable in `RestoreCtx` to indicate whether paretheses should be added when restoring `Join`. Now complicated cases can be restored. Please take a look if I missed something, or are there better solutions?

File: ast/dml.go [hunk 6]
```diff
@@ -85,7 +85,64 @@ type Join struct {
...
+	}
+	ctx.JoinLevel++
+	if err := n.Left.Restore(ctx); err != nil {
+		return errors.Annotate(err, "An error occurred while restore Join.Left")
+	}
+	ctx.JoinLevel--
+	if n.Right == nil {
+		if ctx.JoinLevel != 0 {
+			ctx.WritePlain(")")
+		}
```
Comment (kennytm) [inline ast/dml.go]: Put these into `defer`.

```go
if ctx.JoinLevel != 0 {
    ctx.WritePlain("(")
    defer ctx.WritePlain(")")
}
```

File: ast/dml.go [hunk 7]
```diff
@@ -85,7 +85,59 @@ type Join struct {
 
 // Restore implements Node interface.
 func (n *Join) Restore(ctx *RestoreCtx) error {
-	return errors.New("Not implemented")
+	if ctx.JoinLevel != 0 {
+		ctx.WritePlain("(")
+		defer ctx.WritePlain(")")
+	}
+	ctx.JoinLevel++
```
Comment (leoppro) [inline ast/dml.go]: could we use type asserts to decide whether to add brackets

File: ast/dml_test.go [hunk 8]
```diff
@@ -139,3 +139,36 @@ func (tc *testDMLSuite) TestLimitRestore(c *C) {
 	RunNodeRestoreTest(c, testCases, "SELECT 1 %s", extractNodeFunc)
 }
 
+func (tc *testDMLSuite) TestTableSourceRestore(c *C) {
+	testCases := []NodeRestoreTestCase{
+		{"tbl", "`tbl`"},
+		{"tbl as t", "`tbl` AS `t`"},
+	}
```
[the remark this answers is not in the document]
Comment (exialin) [reply ast/dml_test.go]: @leoppro This is the case for statements like `select * from (select * from a) as t` and `select * from (select * from a union select * from b) as t`. But `SelectStmt` and `UnionStmt` haven't been implemented yet, so I didn't figure out how to test these cases. Do you have any ideas?

[the remark this answers is not in the document]
Comment (leoppro) [reply ast/dml_test.go]: add a `TODO` comment will be nice

File: ast/dml.go (hunk 7 above)
Comment (exialin) [reply ast/dml.go]: Actually I tried as you suggest, but it is hard to handle the following case. For example, based on current parser, the ASTs of `select * from t1, t2` and `select * from (t1) join t2` are the same:
```
*ast.SelectStmt	&{{{{select * from t1,t2}}} {[]} 0xc00000f590 false 0xc00000a720 <nil> 0xc00000f5c0 <nil> <nil> [] <nil> <nil> none [] false false}
  *ast.TableRefsClause	&{{} 0xc000148100}
    *ast.Join	&{{} {[]} 0xc000148080 0xc000052380 1 <nil> [] false false}
      *ast.Join	&{{} {[]} 0xc000052340 <nil> 0 <nil> [] false false}
        *ast.TableSource	&{{} 0xc000146090 }
          *ast.TableName	&{{} {[]}  t1 <nil> <nil> []}
      *ast.TableSource	&{{} 0xc000146120 }
        *ast.TableName	&{{} {[]}  t2 <nil> <nil> []}
  *ast.FieldList	&{{} [0xc000056660]}
    *ast.SelectField	&{{} 7 0xc0000a4d20 <nil>  false}
```
But the AST of `select * from (t1) join (t2)` is different:
```
...
    *ast.Join	&{{} {[]} 0xc000164080 0xc000164100 1 <nil> [] false false}
      *ast.Join	&{{} {[]} 0xc0000a2300 <nil> 0 <nil> [] false false}
        *ast.TableSource	&{{} 0xc000162090 }
          *ast.TableName	&{{} {[]}  t1 <nil> <nil> []}
      *ast.Join	&{{} {[]} 0xc0000a2340 <nil> 0 <nil> [] false false}
        *ast.TableSource	&{{} 0xc000162120 }
          *ast.TableName	&{{} {[]}  t2 <nil> <nil> []}
  ...
```
It is natural to find that `Join`s under the top `Join` should be wrapped with parentheses, and we don't have to think about other cases. If not, we have to consider the actual type of `TableSource`, which could be a normal table, a `SelectStmt` or a `UnionStmt`. I think that is more complicated.

Comment (leoppro) [reply ast/dml.go]: ok, detailed explanation, good job!

**Pull request merged**